### ISSUE TITLE
Per-Layer Dropout Rate Configuration [LayerSkip]

### ISSUE THREAD

**Issue by mostafaelhoushi** [facebookresearch/fairseq2#640]
**Describe the solution you would like:**
Would like to enable configuration of a different layer dropout rate for each layer.

**Describe the alternatives you have considered:**
Currently, layer dropout is implemented in fairseq2 as a scalar probability for all layers (check [here](https://github.com/search?q=repo%3Afacebookresearch%2Ffairseq2+layerdrop&type=code)).
We can follow an implementation similar to this [PR in torchtune](https://github.com/pytorch/torchtune/pull/1076) to support linear, exponential, or step configurations for increasing dropout rate acorss layers.

**Additional Context:**
This will enable implementing: 
- [Progressive Layer Dropping](https://proceedings.neurips.cc/paper/2020/hash/a1140a3d0df1c81e24ae954d935e8926-Abstract.html): that claims to increase accuracy and speed of training if dropout rate increases across layers linearly
- [LayerSkip](https://arxiv.org/abs/2404.16710): that claims to increase accuracy of early exit layers if dropout rate increaes linearly or exponentially across layers

**Comment by mostafaelhoushi** [comment]
The feature (along with other features) was implemented on top of an old fork here: https://github.com/YJYJLee/fairseq2/pull/2
What we can do is:
- The code changes can be ported to the latest fairseq2 main branch.
- Add a couple of unit test cases (similar to testcases [here](https://github.com/mostafaelhoushi/torchtune/blob/layerskip/tests/torchtune/modules/test_layer_dropout.py))

**Comment by HimanshuJanbandhu** [comment]
Hi @mostafaelhoushi, 
I would like to work on this.
